### ISSUE TITLE
CI emits NumPy DeprecationWarning: `__array_wrap__` must accept context and return_scalar arguments

### ISSUE THREAD

**Issue by albertvillanova** [huggingface/trl#6848]
### What

CI emits new `DeprecationWarning`s from NumPy in the GRPO, RLOO and Distillation trainer tests:

```
tests/test_distillation_trainer.py: 24 warnings
tests/test_grpo_trainer.py: 51 warnings
tests/test_rloo_trainer.py: 51 warnings
  /__w/trl/trl/.venv/lib/python3.12/site-packages/numpy/_core/fromnumeric.py:45: DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
    return conv.wrap(result, to_scalar=False)
```

CI run: https://github.com/huggingface/trl/actions/runs/32414561958/job/96572632322

Affected test files:

- `tests/test_grpo_trainer.py` (51 warnings)
- `tests/test_rloo_trainer.py` (51 warnings)
- `tests/test_distillation_trainer.py` (24 warnings)

The tests themselves pass; for now this is warning noise only.

### When it started

First seen in the run for #6619 (`47d20ec6`), and absent from the run for the preceding commit `88b99c2c`: https://github.com/huggingface/trl/actions/runs/32397964455 (4 warnings, none from NumPy).

### Next steps

Investigate the root cause, then decide whether anything is actionable on the TRL side and/or should be reported upstream.

**Comment by albertvillanova** [comment]
## Root cause pinned down

### What triggers it in our CI

Bisected over the CI runs on `main` between `1701d3e` (the runner change, #6741) and `47d20ec6`, grepping each run log for `array_wrap`:

| commit | run | `__array_wrap__` warnings |
| --- | --- | --- |
| `1701d3eb` (#6741) | [32366392208](https://github.com/huggingface/trl/actions/runs/32366392208) | 0 |
| `3f738b90` (#6744) | [32366421880](https://github.com/huggingface/trl/actions/runs/32366421880) | 0 |
| `29b75df4` (#6792) | [32369659020](https://github.com/huggingface/trl/actions/runs/32369659020) | 0 |
| `4d37e9e2` (#6599) | [32394938389](https://github.com/huggingface/trl/actions/runs/32394938389) | 0 |
| `f32c2f6f` (#6763) | [32396277950](https://github.com/huggingface/trl/actions/runs/32396277950) | 0 |
| `040d69df` (#6570) | [32396451216](https://github.com/huggingface/trl/actions/runs/32396451216) | 0 |
| `88b99c2c` (#6838) | [32397964455](https://github.com/huggingface/trl/actions/runs/32397964455) | 0 |
| `47d20ec6` (#6619) | [32414561958](https://github.com/huggingface/trl/actions/runs/32414561958) | **126** |

So this is not the runner change: #6619 added `trl-internal-testing/tiny-InternVLForConditionalGeneration` to the VLM `model_id` parametrization of exactly the three affected files, and the warnings come from the three new test cases it created:

- `tests/test_grpo_trainer.py::TestGRPOTrainerVLM::test_train_vlm[trl-internal-testing/tiny-InternVLForConditionalGeneration]`
- `tests/test_rloo_trainer.py::TestRLOOTrainerVLM::test_train_vlm[trl-internal-testing/tiny-InternVLForConditionalGeneration]`
- `tests/test_distillation_trainer.py::TestDistillationTrainerVLM::test_train_vlm[trl-internal-testing/tiny-InternVLForConditionalGeneration]`

The warning is emitted once per `InternVLProcessor.__call__` that receives images, hence the 51 / 51 / 24 counts (GRPO and RLOO run the same number of processor calls).

### The chain, at the versions CI installs

Checked against the exact pins of that job (`transformers==5.15.1`, `torch==2.13.0`, `numpy==2.5.2`):

1. The InternVL image processor returns `num_patches` inside a `BatchFeature` built with `tensor_type=return_tensors` — `models/got_ocr2/image_processing_got_ocr2.py:267` ([v5.15.1](https://github.com/huggingface/transformers/blob/v5.15.1/src/transformers/models/got_ocr2/image_processing_got_ocr2.py#L267)). The trainers call the processor with `return_tensors="pt"`, so what was a `list[int]` becomes a **`torch.Tensor`**.

2. `InternVLProcessor.__call__` then passes that tensor to NumPy — `models/internvl/processing_internvl.py:110` ([v5.15.1](https://github.com/huggingface/transformers/blob/v5.15.1/src/transformers/models/internvl/processing_internvl.py#L110)):

   ```python
   if images is not None:
       image_num_patches_indices = np.cumsum(model_inputs.pop("num_patches"))
   ```

3. `np.cumsum` on a non-ndarray goes through `_wrapfunc`, which tries `obj.cumsum(axis=…, dtype=…, out=…)`. `Tensor.cumsum` rejects those kwargs with a `TypeError`, so NumPy falls back to `_wrapit` and finishes with `conv.wrap(result, to_scalar=False)` — `numpy/_core/fromnumeric.py:45`, exactly the location reported in CI.

4. `conv.wrap` calls `type(obj).__array_wrap__`, and `torch.Tensor.__array_wrap__(self, array)` still has the pre-NumPy-2.0 two-argument signature (`torch/_tensor.py:1251` at [v2.13.0](https://github.com/pytorch/pytorch/blob/v2.13.0/torch/_tensor.py#L1251), unchanged on `main`) → NumPy emits the deprecation.

Minimal reproduction of steps 3–4, no GPU needed:

```python
>>> import numpy as np, torch, warnings
>>> warnings.simplefilter("error")
>>> np.cumsum(torch.tensor([1, 2, 3]))
DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
```

### Nothing actionable on the TRL side

- We never pass a tensor to NumPy in the code paths involved. The only NumPy calls in the main code are `np.max([t.shape for t in tensors], 0)` in `pad()` (`trl/trainer/utils.py`), which takes `torch.Size` tuples and goes through `_wrapreduction`, not `_wrapfunc`; `np.cumsum` in `trl/data_utils.py`, on Python lists during dataset packing; and `np.array(v)` on lists of `list`s in the vLLM path of the three trainers.
- The `num_patches` tensor never reaches us either: `InternVLProcessor.__call__` pops it before returning.
- Passing `return_tensors=None` to the processor to keep `num_patches` a list is not an option — the trainers need `"pt"` — and the deprecation would still be latent for any user of `InternVLProcessor` with `return_tensors="pt"`.

### Proposed solution

Two upstream defects stack here, and either one alone fixes the warning:

- **`transformers`** — `np.cumsum` on a torch tensor is gratuitous; a `torch.cumsum` (or a cumsum over the list before it is tensorized) removes the NumPy round-trip entirely. This is the actionable report, and it is not specific to our tests: any user calling `InternVLProcessor` with images and `return_tensors="pt"` hits it. No existing issue found.
- **`torch`** — the legacy `__array_wrap__` signature: [pytorch#180657](https://github.com/pytorch/pytorch/issues/180657) (closed as completed, but the code was never actually changed), open PR [pytorch#187250](https://github.com/pytorch/pytorch/pull/187250), plus three closed-unmerged attempts ([#180748](https://github.com/pytorch/pytorch/pull/180748), [#181734](https://github.com/pytorch/pytorch/pull/181734), [#187190](https://github.com/pytorch/pytorch/pull/187190)).

### Next steps

- [ ] Report upstream to `huggingface/transformers`.
- [ ] Decide what to do on the TRL side in the meantime.

**Comment by albertvillanova** [comment]
## Upstream status

Reported upstream: https://github.com/huggingface/transformers/issues/48177

**Not fixed yet.** `np.cumsum(model_inputs.pop("num_patches"))` is identical in the latest release (`v5.15.1`, the version CI installs) and in current `main` — `src/transformers/models/internvl/processing_internvl.py:110` in both.

What I asked for upstream:

1. drop the NumPy round-trip in `InternVLProcessor.__call__` — keep the cumulative sum in torch, or take it over the values before they are tensorized. `image_num_patches_indices` is only ever indexed to slice `pixel_values`, so nothing else changes;
2. check two other processors with the same pattern, where the operand also comes out of a `BatchFeature` and would be a tensor with `return_tensors="pt"`: `models/minicpmv4_6/processing_minicpmv4_6.py:151` and `models/kimi_k25/processing_kimi_k25.py:79` (`models/got_ocr2/processing_got_ocr2.py:237` is fine — it sums a plain Python list).

I offered to send the PR.

The report is not only about our tests: any user calling `InternVLProcessor` with images and `return_tensors="pt"` hits it. It is normally invisible because `DeprecationWarning`s are hidden by default, which is why it surfaces in test suites first.

**No new report for the torch side.** `torch.Tensor.__array_wrap__` keeping the pre-NumPy-2.0 signature is already filed as [pytorch#180657](https://github.com/pytorch/pytorch/issues/180657), and I would not count on it landing soon: the issue is closed as completed although the code was never changed, three PRs have been opened and closed unmerged ([#180748](https://github.com/pytorch/pytorch/pull/180748), [#181734](https://github.com/pytorch/pytorch/pull/181734), [#187190](https://github.com/pytorch/pytorch/pull/187190)), and the currently open one, [#187250](https://github.com/pytorch/pytorch/pull/187250), has been waiting since June. Confirmed still unchanged on `main` (`torch/_tensor.py:1261`). So the transformers fix is the realistic path.

Still to decide: what we do on the TRL side in the meantime.

**Comment by albertvillanova** [comment]
Nothing is actionable on our side, and the fix depends on `transformers` (or, less likely, on `torch`). Meanwhile, I propose to silence it with a `filterwarnings` entry in `pyproject.toml`, documented and linked like the other upstream warning filters, so it can be removed once a fixed `transformers` release is available in CI.

**Comment by albertvillanova** [comment]
I've opened:
- #6850
